src/runtime/storageNG/direct-store.ts

```diff
@@ -0,0 +1,241 @@
+/**
+ * @license
+ * Copyright (c) 2019 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {CRDTModel, CRDTTypeRecord, CRDTChange, ChangeType, CRDTError} from '../crdt/crdt.js';
+import {Type} from '../type.js';
+import {Exists, Driver, DriverFactory} from './drivers/driver-factory.js';
+import {StorageKey} from './storage-key.js';
+import {ActiveStore, ProxyCallback, StorageMode, ProxyMessageType, ProxyMessage} from './store-interface.js';
+
+export enum DirectStoreState {Idle = 'Idle', AwaitingResponse = 'AwaitingResponse', AwaitingResponseDirty = 'AwaitingResponseDirty', AwaitingDriverModel = 'AwaitingDriverModel'}
+
+export class DirectStore<T extends CRDTTypeRecord> extends ActiveStore<T> {
+  localModel: CRDTModel<T>;
+  callbacks = new Map<number, ProxyCallback<T>>();
+  driver: Driver<T['data']>;
+  private nextCallbackID = 1;
+  private version = 0;
+  private pendingException: Error | null = null;
+  private pendingResolves: Function[] = [];
+  private pendingRejects: Function[] = [];
+  private pendingDriverModels: {model: T['data']; version: number;}[] = [];
+  private state: DirectStoreState = DirectStoreState.Idle;
+  /*
+   * This class should only ever be constructed via the static construct method
+   */
+  private constructor(storageKey: StorageKey, exists: Exists, type: Type, mode: StorageMode, modelConstructor: new () => CRDTModel<T>) {
+    super(storageKey, exists, type, mode, modelConstructor);
+  }
+
+  async idle() {
+    if (this.pendingException) {
+      return Promise.reject(this.pendingException);
+    }
+    if (this.state === DirectStoreState.Idle) {
+      return Promise.resolve();
+    }
+    return new Promise<void>((resolve, reject) => {
+      this.pendingResolves.push(resolve);
+      this.pendingRejects.push(reject);
+    });
+  }
+
+  private setState(state: DirectStoreState) {
+    this.state = state;
+    if (state === DirectStoreState.Idle) {
+      // If we are already idle, this won't notify external parties.
+      this.notifyIdle();
+    }
+  }
+  
+  private notifyIdle() {
+    if (this.pendingException) {
+      // this is termination.
+      this.pendingRejects.forEach(reject => reject(this.pendingException));
```

Comment:
Do you want to clear the list of pendingRejects after this? Or is that what the "this is termination" comment is referring to?

Reply:
Yeah.